depeche_db/experimental/threaded_executor.py

```diff
@@ -0,0 +1,152 @@
+            str, List[HandlerRegistration]
+        ] = _collections.defaultdict(list)
+        self.stimulation_interval = stimulation_interval
+        self.keep_running = True
+        self.handler_queues = {}
+        self.stimulator_thread = _threading.Thread(target=self._stimulate, daemon=True)
+        self.handler_threads = []
+        self.listener = None
+        if not disable_signals:
+            _signal.signal(_signal.SIGINT, lambda *_: self._stop())
+            _signal.signal(_signal.SIGTERM, lambda *_: self._stop())
+
+    def register(self, handler: RunOnNotification, concurrency: int = 1):
+        """
+        Registers a handler to be run on notifications.
+
+        Args:
+            handler: Handler to register
+        """
+        self.channel_register[handler.notification_channel].append(
+            HandlerRegistration(handler=handler, concurrency=concurrency)
+        )
+        return handler
+
+    def _stop(self):
+        self.keep_running = False
+        if self.listener is not None:
+            self.listener.stop()
+
+        for registrations in self.channel_register.values():
+            for registration in registrations:
+                registration.handler.stop()
+
+    def run(self):
+        """Runs the executor."""
+        self.listener = PgNotificationListener(
+            dsn=self._db_dsn,
+            channels=list(self.channel_register),
+            ignore_payload=False,
+        )
+
+        for registrations in self.channel_register.values():
+            for registration in registrations:
+                self.handler_queues[registration.id] = UniqueQueue()
+                for _ in range(registration.concurrency):
+                    self.handler_threads.append(
+                        _threading.Thread(
+                            target=self._run_handler(registration),
+                            daemon=True,
+                        )
+                    )
+        for thread in self.handler_threads:
+            thread.start()
+        self.stimulator_thread.start()
+        self.listener.start()
+
+        for notification in self.listener.messages():
+            for registration in self.channel_register[notification.channel]:
+                if registration.handler.interested_in_notification(
+                    notification.payload
+                ):
+                    registration.handler.take_notification_hint(notification.payload)
+                    self.handler_queues[registration.id].put(registration.id)
+
+        for thread in self.handler_threads:
+            thread.join()
+        self.stimulator_thread.join()
+
+    def _run_handler(self, registration: HandlerRegistration):
+        def run_handler():
+            while self.keep_running:
+                try:
+                    self.handler_queues[registration.id].get(timeout=0.5)
+                    result = registration.handler.run(budget=FixedTimeBudget(seconds=3))
```

Comment:
The `run` method of this `RunOnNotification` instance needs to be thread-safe (at least if `concurrency` is above 1 for that registration...